### PR TITLE
Feature/number input formatting

### PR DESCRIPTION
**Pull request by RafaelVidaurre**
**Please make sure the following boxes are checked before submitting your Pull Request:**

- [X] I linked the issues related to this PR in its description (if any).
- [X] If this PR adds new code that is not going to change in the near future, it includes unit tests to cover it.

# This PR

Fixes Number input to support decimals and correct formatting.

Implements `formikProxy` as an adapter for primitive inputs. Ideally we want to migrate al of our inputs to this pattern:

- We have a primitive input that is agnostic to its context
- We export it wrapped with the `formikProxy` for use with Formik forms.

This allows us decouple our components from the Formik context. The `formikProxy` is a simple Hoc that works as an adapter for fields to work with Formik.

NOTE: This PR does not propose an export/import pattern, I think we still need to discuss how we should deal with our component modules to be consistent everywhere.

### CODE REVIEW

File: packages/polymath-ui/src/components/inputs/CurrencySelect/CurrencySelect.js [hunk 1]
```diff
@@ -1,35 +1,39 @@
...
   handleChange = value => {
-    const {
-      field,
-      form: { setFieldValue, setFieldTouched },
-    } = this.props;
-    setFieldValue(field.name, value);
-    setFieldTouched(field.name, true);
+    const { onChange, onBlur } = this.props;
+    onChange(value);
+    onBlur();
```
Comment (monitz87) [inline packages/polymath-ui/src/components/inputs/CurrencySelect/CurrencySelect.js]: Why is the component blurred after every change?

File: packages/polymath-ui/src/components/inputs/BaseInput/index.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ const Unit = styled.span`
 export default ({ unit, ...props }) => {
   return (
     <Container unit={unit}>
-      <Input unit={unit} {...props} />
+      <Input data-testid="base-input" unit={unit} {...props} />
```
Comment (monitz87) [inline packages/polymath-ui/src/components/inputs/BaseInput/index.js]: What is the purpose of `data-testid`?

File: packages/polymath-ui/src/components/inputs/NumberInput/index.js [hunk 3]
```diff
@@ -1,52 +1,156 @@
...
+  onChange?: value => void,
+  onBlur?: () => void,
+|};
+
+type State = {|
+  oldValue: null,
+  displayValue: string,
+|};
+
+// Any state which is valid for desplaying in the input. If the new value of
```
Comment (monitz87) [inline packages/polymath-ui/src/components/inputs/NumberInput/index.js]:
```suggestion
// Any state which is valid for displaying in the input. If the new value of
```

File: packages/polymath-ui/src/components/inputs/NumberInput/index.js [hunk 4]
```diff
@@ -1,52 +1,156 @@
...
+// the input doesn't match this, it will rollback to a previous state
+const displayValueRegex = /(^[\d,]*\.?$)|(^\.$)|(^\.[0,]|[1-9,][\d,]*$)|(^[\d,]*\.[\d,]*$)/;
+
+// States that cannot be formatted
+const endsWithZeroInDecimalsRegex = /^[\d,]*\.[,\d]*[0,]$/;
+const pendingDot = /^[\d,]*\.$/;
```
Comment (monitz87) [inline packages/polymath-ui/src/components/inputs/NumberInput/index.js]:
```suggestion
const pendingDotRegex = /^[\d,]*\.$/;
```

File: packages/polymath-ui/src/components/inputs/NumberInput/index.js [hunk 5]
```diff
@@ -1,52 +1,156 @@
...
+   *
+   * @static
+   * @param displayValue The value in the input that the user sees
+   */
+  static isInIntermediateState(displayValue: string) {
+    // const startsWithDot = startsWithDotRegex.test(displayValue);
+    const endsWithZeroInDecimals = endsWithZeroInDecimalsRegex.test(
+      displayValue
+    );
+    const pendingDor = pendingDot.test(displayValue);
```
Comment (monitz87) [inline packages/polymath-ui/src/components/inputs/NumberInput/index.js]:
```suggestion
    const hasPendingDot = pendingDotRegex.test(displayValue);
```

File: packages/polymath-ui/src/components/inputs/NumberInput/index.js [hunk 6]
```diff
@@ -1,52 +1,156 @@
...
+
+    return endsWithZeroInDecimals || pendingDor;
```
Comment (monitz87) [inline packages/polymath-ui/src/components/inputs/NumberInput/index.js]:
```suggestion
    return endsWithZeroInDecimals || hasPendingDot;
```

File: packages/polymath-ui/src/components/inputs/NumberInput/index.js (hunk 4 above)
Comment (RafaelVidaurre) [reply packages/polymath-ui/src/components/inputs/NumberInput/index.js]: 👍

**Comment by RafaelVidaurre** [comment]
Merging this without updating the other inputs for now, sadly time constraints are in the way 🆗

**Pull request merged**